modules/markup/markdown/toc.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package markdown
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/unknwon/i18n"
+	"github.com/yuin/goldmark/ast"
+)
+
+func createTOCNode(toc []Header, lang string) ast.Node {
+	details := NewDetails()
+	summary := NewSummary()
+
+	summary.AppendChild(summary, ast.NewString([]byte(i18n.Tr(lang, "toc"))))
+	details.AppendChild(details, summary)
+	ul := ast.NewList('-')
+	details.AppendChild(details, ul)
+	currentLevel := 6
+	for _, header := range toc {
+		if header.Level < currentLevel {
+			currentLevel = header.Level
+		}
+	}
+	for _, header := range toc {
+		if header.Level < currentLevel {
+			ul = ul.Parent().(*ast.List)
+		} else if header.Level > currentLevel {
+			for currentLevel < header.Level {
+				newL := ast.NewList('-')
+				ul.AppendChild(ul, newL)
+				currentLevel++
```

Comment:
Are levels always distanced by 1? (e.g. can't we have level 5 inside level 3 inside level1?)

Reply:
That's what this code is handling.